
\myreference
{ilinkin2010}
{Ilinkin,I. et al. (2010) 
Multiple structure alignment and consensus identification for proteins. 
\emph{BMC Bioinformatics}, \textbf{11}, 71.}
{https://doi.org/10.1186/1471-2105-11-71}

\myreference
{midlik2019}
{Midlik,A. et al. (2019) 
Automated family-wide annotation of secondary structure elements. 
\emph{Methods Mol Biol}, \textbf{1958}, 47--71.}
{https://doi.org/10.1007/978-1-4939-9161-7_3}

\myreference
{midlik2021}
{Midlik,A. et al. (2021) 
Uncovering of cytochrome P450 anatomy by SecStrAnnotator. 
\emph{Sci Rep}, \textbf{11}, 12345.}
{https://doi.org/10.1038/s41598-021-91494-8}

\myreference
{needleman1970}
{Needleman,S.B. and Wunsch,C.D. (1970)
A general method applicable to the search for similarities in the amino acid sequence of two proteins.
\emph{J Mol Biol}, \textbf{48}, 443--453.}
{https://doi.org/10.1016/0022-2836(70)90057-4}

\myreference
{schrodinger}
{Schr√∂dinger, LLC. The PyMOL Molecular Graphics System, Version 2.3}
{https://pymol.org/2/}

\myreference
{shindyalov1998}
{Shindyalov,I.N. and Bourne,P.E. (1998) 
Protein structure alignment by incremental combinatorial extension (CE) of the optimal path. 
\emph{Protein Eng}, \textbf{11}, 739--747.}
{https://doi.org/10.1093/protein/11.9.739}

\myreference
{sillitoe2021}
{Sillitoe,I. et al. (2021) 
CATH: increased structural coverage of functional space. 
\emph{Nucleic Acids Res}, \textbf{49}, D266--D273.}
{https://doi.org/10.1093/nar/gkaa1079}

\myreference
{wwpdb2019}
{wwPDB consortium. (2019) 
Protein Data Bank: the single global archive for 3D macromolecular structure data. 
\emph{Nucleic Acids Res}, \textbf{47}, D520-D528.}
{https://doi.org/10.1093/nar/gky949}

